### PR TITLE
core part of encryption 2.0

### PR DESCRIPTION
**Pull request by schiessle**
this is the core part of encryption 2.0. It also contains a dummy encryption module which allows you to test the core functionallity without modifiyng/encrypting the data.

To enable the new encryption first enable the "dummy encryption module" app and than do to the admin settings to enable server-side encryption.

Known issue:
- [ ] doesn't work for large files, I will work on this issue independently.

cc @DeepDiver1975 @karlitschek @th3fallen

### CODE REVIEW

**Comment by karlitschek** [comment]
as discussed :+1:

**Comment by schiesbn** [comment]
@jknockaert I think the problem with large files is related to my port of your PR "make encryption stream wrapper seekable" https://github.com/owncloud/core/pull/14383

On master everything works fine so I assume I introduced the problem while porting it to the new encryption. Maybe it is also not 100% compatible with the new encryprion. Would be great if you could also have a look at it if you have some time. I have the feeling there is something wrong with the calculation of the position because we don't only lose data at the end of large files but also in between.

File: tests/lib/encryption/managertest.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		$em->expects($this->any())->method('getId')->willReturn(0);
+		$em->expects($this->any())->method('getDisplayName')->willReturn('TestDummyModule0');
+		$m = new Manager($config);
+		$m->registerEncryptionModule($em);
+		$this->assertSame(1, count($m->getEncryptionModules()));
+		$en0 = $m->getEncryptionModule(0);
+		$this->assertEquals(0, $en0->getId());
+	}
+
+	public function XtestGetDefaultEncryptionModule() {
```
Comment (DeepDiver1975) [inline tests/lib/encryption/managertest.php]: why is this not being executed? future test? please add comment - THX

**Comment by jknockaert** [comment]
@schiesbn I remember that when I developed the code that a specific problem happened at around 8GB; I solved that problem indeed. I could have a look at the code next Sunday.

File: lib/private/files/stream/encryption.php [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OC\Files\Stream;
+
+use Icewind\Streams\Wrapper;
+use OC\Encryption\Exceptions\EncryptionHeaderKeyExistsException;
+
+class Encryption extends Wrapper {
```
Comment (DeepDiver1975) [inline lib/private/files/stream/encryption.php]: @schiesbn is there a unit test on this class? just asking ... :speak_no_evil:

File: lib/private/files/fileinfo.php [hunk 3]
```diff
@@ -151,6 +151,13 @@ public function isEncrypted() {
 	/**
 	 * @return int
 	 */
+	public function getUnencryptedSize() {
```
Comment (DeepDiver1975) [inline lib/private/files/fileinfo.php]: we wanted to kill this right? nothing on this pr - but we shall not forget - THX

File: lib/private/encryption/util.php [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OC\Encryption;
+
+use OC\Encryption\Exceptions\EncryptionHeaderToLargeException;
+use OC\Encryption\Exceptions\EncryptionHeaderKeyExistsException;
+use OCP\Encryption\IEncryptionModule;
+
+class Util {
```
Comment (DeepDiver1975) [inline lib/private/encryption/util.php]: unit test?

Comment (schiesbn) [inline lib/private/encryption/util.php]: -> tests/lib/encryption/utiltest.php

File: tests/lib/encryption/managertest.php (hunk 1 above)
Comment (schiesbn) [inline tests/lib/encryption/managertest.php]: enabled

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by jknockaert** [comment]
@schiesbn Provided that the PR is merged, does it mean that the big-file problem is solved and no review is needed anymore? I don't really have an overview of the encryption 2.0 development.

**Comment by DeepDiver1975** [comment]
> Known issue:
> 
>     doesn't work for large files, I will work on this issue independently.

@jknockaert the big-file issue is still open - we will address this in another pr.

**Comment by jknockaert** [comment]
@DeepDiver1975 Thanks for clarifying.

**Comment by schiesbn** [comment]
@jknockaert would be awesome if you would find some time to look into it! Thanks a lot!

**Comment by jknockaert** [comment]
@schiesbn OK, I had a look at lib/private/files/storage/wrapper/encryption.php. There's more than a couple of issues that I want to raise. I'm not sure what would be the right format to do so. But my main concern is that you did not port the private readCache() function. There may be reasons not to do so, but the tests at the beginning of that function do have a reason. Just imagine the scenario that you open a file, read 1 character, and read again 1 character. My readCache() makes sure that upon the second read the same cached block is used. In your implementation of stream_read any read starts by loading the next block. It may be that this scenario doesn't happen in the wild. As for the write issues, I am not sure how they happen, but the stream_write at L291 is not something you ported from my implementation. As long as the block isn't full, it shouldn't be written, unless the file is closed, or there is an fseek. The private functions readCache() and flush() play a central role in my implementation and are called at many places, not limited to read and write but also in seek, close and (the public function) flush.

**Comment by jknockaert** [comment]
@schiesbn Further looking at  lib/private/files/stream/encryption.php I noticed that you introduced a write in the else-clause at L309. That's not correct, because when the if condition is not met, nothing should be written (which is handled in my code by setting data='', a length 0 will then be returned which signals that nothing was written). Generally I expect your implementation to fail on multiple points when the encrypted stream is nonseekable, a case which should be considered: https://github.com/owncloud/core/pull/8390#issuecomment-41647795. The structure of my implementation is such that when leaving the current block the private function flush() is called, which writes the block (if the writeflag is set) and empties the cache ($decrypted in your code). Leaving the block may happen in read, write, seek or close, so it should be invoked in each of these functions at the appropriate place. When reading or writing on the current block, readCache is called which basically checks if the current block is already in the cache, and if not, it is loaded. There is some additional logic for handling end-of-file writes correctly. This mechanism ensures correct handling of nonseekable encrypted streams, as the pointer in the nonseekable stream moves upon each read (you can't rewind or seek). In the comments of the different incarnations of my implementation there's some more documentation of the chosen approach.

**Comment by jknockaert** [comment]
@schiesbn Further checking, I found that at L303 you mix the size of the encrypted and the unencrypted stream in the max argument; that is not correct.

**Comment by jknockaert** [comment]
I created #15303 based on my comments above

**Pull request merged**